src/update.py

```diff
@@ -27,56 +27,99 @@
-    table_name = pipeline_name.replace("_", "-")
-    if args.only and not table_name in args.only.split(","):
-        continue
-    if args.exclude and table_name in args.exclude.split(","):
-        continue
-    pipeline_chain = DataPipeline.load(pipeline_name)
-    show_progress = not args.no_progress
-    pipeline_output = pipeline_chain.run(
-        pipeline_name, verify=args.verify, process_count=args.process_count, progress=show_progress
+
+def main(
+    output_folder: Path,
+    verify: str = None,
+    only: List[str] = None,
+    exclude: List[str] = None,
+    process_count: int = cpu_count(),
+    show_progress: bool = True,
+) -> None:
+    """
+    Executes the data pipelines and places all outputs into `output_folder`. This is typically
+    followed by publishing of the contents of the output folder to a server.
+
+    Args:
+        output_folder: Root folder where snapshot, intermediate and tables will be placed.
+        verify: Run anomaly detection on the outputs using this strategy. Value must be one of:
+            - None: (default) perform no anomaly detection
+            - "simple": perform only fast anomaly detection
+            - "full": perform exhaustive anomaly detection (can be very slow)
+        only: If provided, only pipelines with a name appearing in this list will be run.
+        exclude: If provided, pipelines with a name appearing in this list will not be run.
+        process_count: Maximum number of processes to use during the data pipeline execution.
+        show_progress: Display progress for the execution of individual DataSources within this
+            pipeline.
+    """
+
+    assert not (
+        only is not None and exclude is not None
+    ), "--only and --exclude options cannot be used simultaneously"
+
+    # Ensure that there is an output folder to put the data in
+    (output_folder / "snapshot").mkdir(parents=True, exist_ok=True)
+    (output_folder / "intermediate").mkdir(parents=True, exist_ok=True)
+    (output_folder / "tables").mkdir(parents=True, exist_ok=True)
+
+    # A pipeline chain is any subfolder not starting with "_" in the pipelines folder
+    all_pipeline_chains = []
+    for item in (ROOT / "src" / "pipelines").iterdir():
+        if not item.name.startswith("_") and not item.is_file():
+            all_pipeline_chains.append(item.name)
+
+    # Verify that all of the provided pipeline names exist as pipelines
+    only = only.split(",") if only is not None else []
+    exclude = exclude.split(",") if exclude is not None else []
+    for pipeline_name in only + exclude:
+        assert pipeline_name in all_pipeline_chains, f'"{pipeline_name}" pipeline does not exist'
+
+    # Run all the pipelines and place their outputs into the output folder
+    # The output name for each pipeline chain will be the name of the directory that the chain is in
+    for pipeline_name in all_pipeline_chains:
+        table_name = pipeline_name.replace("_", "-")
```

Comment:
Just for the sake of argument--is this replacement necessary?

Reply:
URL names typically have dashes, but I can't have dashes in Python package names. That's literally the only reason, otherwise I would have named my Python packages using dashes.